plugin/src/main/java/com/google/errorprone/xplat/checker/UnnecessaryConcurrentHashMap.java

```diff
@@ -14,56 +14,197 @@
+
+      if (variable == null) {
+        return Description.NO_MATCH;
+      } else if (variable.getKind() == Kind.VARIABLE) {
+        String source = state.getSourceForNode(variable);
+
+        if (source != null && source.contains("<") && !MAP_MATCHER.matches(variable, state)) {
+          fix.addImport("java.util.Map");
+          fix.replace(
+              ((JCTree) variable).getStartPosition(),
+              ((JCTree) variable).getStartPosition() + source.indexOf("<"),
+              "Map");
+        }
+
+      } else if (variable.getKind() == Kind.ASSIGNMENT) {
+        Optional<Description> desc = incompatibleInterfaceDesc(variable, state,
+            state.getSourceForNode(variable));
+
+        if (desc.isPresent()) {
+          state.reportMatch(standardDescription(tree, fix.build(), STANDARD_MESSAGE));
+
+          return desc.get();
+        } else {
+          return standardDescription(tree, fix.build(), STANDARD_MESSAGE +
+              " Make sure that this variable is declared with the Map interface.");
+        }
+      }
+      return standardDescription(tree, fix.build(), STANDARD_MESSAGE);
     }
+
     return Description.NO_MATCH;
   }
 
+  /**
+   * Handles the edge case where the ConcurrentHashMap is declared with an interface that is
+   * incompatible with Collections.synchronizedMap on a different line than it is instantiated. This
+   * only handles the case where both are in the same scope. Otherwise, it is on the user to verify
+   * that the variable is declared with a valid interface.
+   */
+  private Optional<Description> incompatibleInterfaceDesc(Tree tree, VisitorState state,
+      String source) {
+
+    String treeSource = state.getSourceForNode(tree);
+
+    if (treeSource == null || source == null || !source.contains("=")) {
+      return Optional.empty();
+    }
+
+    String originName = source.substring(0, treeSource.indexOf("=")).trim();
+
+    VariableTree origin =
+        new TreePathScanner<VariableTree, Void>() {
+          @Override
+          public VariableTree reduce(VariableTree a, VariableTree b) {
+            return a == null ? b : a;
+          }
+
+          @Override
+          public VariableTree visitVariable(VariableTree node, Void unused) {
+            if (node.getName().toString().equals(originName)) {
+              return node;
+            }
+            return super.visitVariable(node, unused);
+          }
+        }.scan(state.getPath().getParentPath().getParentPath().getParentPath(), null);
+
+    if (origin == null) {
+      return Optional.empty();
+    }
+
+    String originSource = state.getSourceForNode(origin);
+
+    if (originSource != null && OTHER_MAP_INTERFACE_MATCHER.matches(origin, state)) {
+
+      return Optional.of(buildDescription(origin)
+          .setMessage("This variable is declared with an interface that is not compatible"
+              + " with Collections.synchronizedMap, which is advised over ConcurrentHashMap.")
```

Comment:
I think more details about exectly why it's not advised to use ConcurrentHashMap would be helpful here.

In the past we've seen that we get a lot less pushback for users on suggestions if they understand the logic of the suggestion.

Reply:
I have updated this message to reflect that it is meant to be given with another error that has a better explanation.